dvc/remote/local/__init__.py

```diff
@@ -572,3 +572,45 @@ def _get_unpacked_dir_names(self, checksums):
             if self.is_dir_checksum(c):
                 unpacked.add(c + self.UNPACKED_DIR_SUFFIX)
         return unpacked
+
+    def _needs_checkout(self, path_info, checksum):
+        # NOTE: In case if path_info is already cached, and cache type is
+        # 'copy' we would like to avoid relinking.
+        # `symlink/hardlink/reflink` is cheap, but if we need `copy` links,
+        # then to save on copying we should only relink if we already have
+        # `hardlink` or a `symlink`, we don't care about reflinks, because
+        # they are indistinguishable from copy anyway.
+
+        if (
+            self._is_cache_copy(path_info)
+            and not self.changed(path_info, {self.PARAM_CHECKSUM: checksum})
```

Comment:
maybe lets move this down below, since is_hardlink and is_symlink are much cheaper operations then accessing the db?

Reply:
`is_...link` assumes existance of given `path_info`, which is not always possible (`test_checkout.TestCheckoutSingleStage`, for example). So I can:
- use `self.changed` first, which covers case when `path_info` does not exists
or
- make one more `self.exists(path_info)` before hardlinks check.

I went with `self.changed` because I though duplicating calls will be bigger sin.
What do you think?